### PR TITLE
feat(mito): merge reader for mito2

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR implements a `MergeReader` to merge multiple sorted batch sources into one and also remove duplications.

The `MergeReader` requires all input source yields sorted and deduplicated Batches. The reader uses a heap to sort and collect batches for each series.

Currently, the reader only removes deleted rows but doesn't filter rows by sequence for simplicity and performance reason. So it is possible to read rows in another write request that is still writing.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #1869
- Requires #2203

### CODE REVIEW

File: src/mito2/src/read/merge.rs [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+use async_trait::async_trait;
+
+use crate::error::Result;
+use crate::read::{Batch, BatchReader, BoxedBatchReader, Source};
+
+/// Reader to merge sorted batches.
+///
+/// The merge reader merges [Batch]es from multiple sources that yields sorted batches.
+/// Batches from each source **must** obey the following rules:
```
Comment (v0y4g3r) [inline src/mito2/src/read/merge.rs:29]:
```suggestion
/// The merge reader merges [Batch]es from multiple sources that yield sorted batches.
```

File: src/mito2/src/read/merge.rs [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            // Safety: node is not empty.
+            if self.nodes.peek().unwrap().primary_key() != current_key {
+                break;
+            }
+            // They have the same primary key, we could take it and try next node.
+            self.take_batch_from_heap().await?;
+        }
+
+        // Merge collected batches.
+        self.batch_merger.merge_batches()
```
Comment (v0y4g3r) [inline src/mito2/src/read/merge.rs:65]: Looks like `MergeReader` will also do merge/sort/dedup when TimeSeriesMemtable has done merge/sort/dedup. Maybe we can find a fast-path when batches from memtable do not overlap with each other.

**Review by v0y4g3r — APPROVED** [review]
LGTM

File: src/mito2/src/read/merge.rs [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+use crate::error::Result;
+use crate::memtable::BoxedBatchIterator;
+use crate::read::{Batch, BatchReader, BoxedBatchReader, Source};
+
+/// Reader to merge sorted batches.
+///
+/// The merge reader merges [Batch]es from multiple sources that yield sorted batches.
+/// 1. Batch is ordered by primary key, time index, sequence desc, op type desc (we can
+/// ignore op type as sequence is already unique).
+/// 2. Batch doesn't have duplicate elements (element with same key).
```
Comment (waynexia) [inline src/mito2/src/read/merge.rs:32]:
```suggestion
/// 2. Batch doesn't have duplicate elements (elements with the same key and time index).
```

**Review by waynexia — APPROVED** [review]

**Pull request merged**